libp2p-kad/src/query.rs

```diff
@@ -0,0 +1,335 @@
+use kad_server::KademliaServerController;
+use kbucket::KBucketsPeerId;
+use libp2p_peerstore::PeerId;
+use multiaddr::{AddrComponent, Multiaddr};
+use protocol;
+use rand;
+use smallvec::SmallVec;
+use std::cmp::Ordering;
+use std::io::Error as IoError;
+use std::mem;
+use std::time::Duration;
+
+/// Interface that the query uses to communicate with the rest of the system.
+pub trait QueryInterface: Clone {
+    /// Returns the peer ID of the local node.
+    fn local_id(&self) -> &PeerId;
+
+    /// Finds the nodes closest to a peer ID.
+    fn kbuckets_find_closest(&self, addr: &PeerId) -> Vec<PeerId>;
+
+    /// Adds new known multiaddrs for the given peer.
+    fn peer_add_addrs<I>(&self, peer: &PeerId, multiaddrs: I, ttl: Duration)
+    where
+        I: Iterator<Item = Multiaddr>;
+
+    /// Returns the level of parallelism wanted for this query.
+    fn parallelism(&self) -> usize;
+
+    /// Duration between two waves of queries.
+    fn cycle_duration(&self) -> Duration;
+
+    /// Attempts to contact the given multiaddress, then calls `and_then` on success. Returns a
+    /// future that contains the output of `and_then`, or an error if we failed to contact the
+    /// remote.
+    // TODO: use HKTB once Rust supports that, to avoid boxing the future
+    fn send<F, FRet>(
+        &self,
+        addr: Multiaddr,
+        and_then: F,
+    ) -> Box<Future<Item = FRet, Error = IoError>>
+    where
+        F: FnOnce(&KademliaServerController) -> FRet + 'static,
+        FRet: 'static;
+}
+
+/// Starts a query for an iterative `FIND_NODE` request.
+#[inline]
+pub fn find_node<'a, I>(
+    query_interface: I,
+    searched_key: PeerId,
+) -> Box<Future<Item = Vec<PeerId>, Error = IoError> + 'a>
+where
+    I: QueryInterface + 'a,
+{
+    query(query_interface, searched_key, 20) // TODO: constant
+}
+
+/// Refreshes a specific bucket by performing an iterative `FIND_NODE` on a random ID of this
+/// bucket.
+///
+/// Returns a dummy no-op future if `bucket_num` is out of range.
+pub fn refresh<'a, I>(
+    query_interface: I,
+    bucket_num: usize,
+) -> Box<Future<Item = (), Error = IoError> + 'a>
+where
+    I: QueryInterface + 'a,
+{
+    let peer_id = match gen_random_id(&query_interface, bucket_num) {
+        Ok(p) => p,
+        Err(()) => return Box::new(future::ok(())),
+    };
+
+    let future = find_node(query_interface, peer_id).map(|_| ());
+    Box::new(future) as Box<_>
+}
+
+// Generates a random `PeerId` that belongs to the given bucket.
+//
+// Returns an error if `bucket_num` is out of range.
+fn gen_random_id<I>(query_interface: &I, bucket_num: usize) -> Result<PeerId, ()>
+where
+    I: ?Sized + QueryInterface,
+{
+    // TODO: there is a design issue here, due to the fact that peer IDs are multihashes and not
+    //       byte arrays
+
+    let my_id = query_interface.local_id();
+    let my_id_len = my_id.as_bytes().len();
+
+    // TODO: this 2 is magic here ; it is the length of the hash of the multihash
```

Comment:
Compute it somehow (so this doesn't break) or at least extract it to a constant

Reply:
Doing it properly requires a big rework of the `rust-multihash` crate, which is why I just put this hack here.